### PR TITLE
feat: add string/for-each

### LINKED ISSUE #855: [RFC]: Add `@stdlib/string/for-each`
### Description

This RFC proposes adding support for invoking a callback for each (visual) character of a string.

Signature:

```javascript
forEach( str, clbk[, thisArg] )
```

where `clbk` (similar to [`@stdlib/utils/for-each`](https://github.com/stdlib-js/stdlib/tree/ee2d245f666240fc94f6bdfcf78d62f1883773a8/lib/node_modules/%40stdlib/utils/for-each)) is provided

- **value**: visual character
- **idx**: starting character index
- **str**: input string

Package: `@stdlib/string/for-each`
Alias: `forEach`

### Related Issues

None.

### Questions

No.

### Other

Will require iterating over grapheme cluster breaks. See [`@stdlib/string/reverse`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/string/reverse) for such an example.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Infinage**
## Description

> What is the purpose of this pull request?

This pull request:

-   Adds support for iterating through every (visual) character of a string, while invoking a specified callback on each.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #855 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/string/for-each/README.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
+<!--
+
+@license Apache-2.0
+
+Copyright (c) 2018 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/README.md]:
```suggestion
Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/README.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+<section class="notes">
+
+## Notes
+
+-   The function differs from traditional string iterations in the following ways:
+
+    -   The function returns the input `str`.
+
+    -   The `clbk` is executed on each _visual_ character as opposed to being executed on each character
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/README.md]:
```suggestion
    -   The `clbk` is executed on each _visual_ character as opposed to being executed on each character.
```

File: lib/node_modules/@stdlib/string/for-each/README.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    8: r
+    9: l
+    10: d
+    11: !
+*/
+```
+
+The invoked `function` is provided three arguments:
+
+-   `value`: visual character
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/README.md]:
```suggestion
-   `value`: visual character.
```

File: lib/node_modules/@stdlib/string/for-each/README.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+-   `index`: starting character index
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/README.md]:
```suggestion
-   `index`: starting character index.
```

File: lib/node_modules/@stdlib/string/for-each/README.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+-   `str`: input string
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/README.md]:
```suggestion
-   `str`: input string.
```

File: lib/node_modules/@stdlib/string/for-each/docs/types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+/*
+* @license Apache-2.0
+*
+* Copyright (c) 2019 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/docs/types/index.d.ts]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/benchmark/benchmark.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/benchmark/benchmark.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/docs/types/test.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
+/*
+* @license Apache-2.0
+*
+* Copyright (c) 2019 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/docs/types/test.ts]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/test/test.js [hunk 9]
```diff
@@ -0,0 +1,178 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/test/test.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/examples/index.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/examples/index.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/examples/index.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+var forEach = require( './../lib' );
+
+/* eslint-disable no-unused-vars */
+function log( value, index, str ) {
+	var repeatStr = value + value;
+	console.log( '%s: %d', index, repeatStr );
+}
+
+// Iterates over each visual character
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/examples/index.js]:
```suggestion
// Iterates over each visual character:
```

File: lib/node_modules/@stdlib/string/for-each/examples/index.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+forEach( 'presidential election', log );
+
+// Iterates over each visual character
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/examples/index.js]:
```suggestion
// Iterates over each visual character:
```

File: lib/node_modules/@stdlib/string/for-each/examples/index.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+forEach( 'Iñtërnâtiônàlizætiøn', log );
+
+// Iterates over each visual character
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/examples/index.js]:
```suggestion
// Iterates over each visual character:
```

File: lib/node_modules/@stdlib/string/for-each/examples/index.js [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+forEach( '🌷🍕', log );
+
+// Iterates over each visual character
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/examples/index.js]:
```suggestion
// Iterates over each visual character:
```

File: lib/node_modules/@stdlib/string/for-each/lib/index.js [hunk 15]
```diff
@@ -0,0 +1,45 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/lib/index.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/lib/main.js [hunk 16]
```diff
@@ -0,0 +1,87 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/lib/main.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/for-each/lib/main.js [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+		throw new TypeError( format( 'invalid argument. Second argument must be a function. Value: `%s`.', clbk ) );
+	}
+	if ( str === '' ) {
+		return '';
+	}
+
+	idx = 0;
+	len = str.length;
+	while ( idx < len ) {
+		// Find out where the next grapheme cluster begins
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/lib/main.js]:
```suggestion
		// Find out where the next grapheme cluster begins:
```

File: lib/node_modules/@stdlib/string/for-each/lib/main.js [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+		brk = nextGraphemeClusterBreak( str, idx );
+
+		// -1 only when end of string is reached
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/lib/main.js]:
```suggestion
		// -1 only when end of string is reached...
```

File: lib/node_modules/@stdlib/string/for-each/lib/main.js [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+		if (brk === -1) {
+			brk = len;
+		}
+
+		// Invoke the clbk on the cluster
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/lib/main.js]:
```suggestion
		// Invoke the clbk on the cluster:
```

File: lib/node_modules/@stdlib/string/for-each/lib/main.js [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+		clbk.call( thisArg, str.substring(idx, brk), idx, str );
+
+		// Hop onto the beginning of the next cluster
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/for-each/lib/main.js]:
```suggestion
		// Hop onto the beginning of the next cluster:
```

**Review by Planeshifter — CHANGES_REQUESTED** [review]

**Review by Planeshifter — APPROVED** [review]
Just made some small documentation changes and updated the copyright years in the license comments. Will merge in once CI passes.

**Comment by Infinage** [comment]
Thank you!

**Pull request merged**